File: binary_search/34.find-first-and-last-position-of-element-in-sorted-array.cpp
/*
 * @lc app=leetcode id=34 lang=cpp
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */

#include "bits/stdc++.h"

using namespace std;


// @lc code=start
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if(nums.empty()){
            return {-1, -1};
        }
        int lower = lower_bound(nums.begin(), nums.end(), target)- nums.begin();
        int upper = upper_bound(nums.begin(), nums.end(), target) - nums.begin() - 1;

        int len = nums.size();

        if(lower == len || nums[lower] != target){
            return vector<int>{-1, -1};
        }

        return {lower, upper};
    }
};
// @lc code=end

File: binary_search/34.find-first-and-last-position-of-element-in-sorted-array.py
#
# @lc app=leetcode id=34 lang=python3
#
# [34] Find First and Last Position of Element in Sorted Array
#

# @lc code=start
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:

        def binarySearchLeft():
            l, r = 0, len(nums)-1
            while l<=r:
                m = l + (r-l)//2
                if nums[m] < target:
                    l =m+1
                else:
                    r = m-1
            return l

        def binarySearchRight():
            l, r = 0, len(nums)-1
            while l <= r:
                m = l + (r-l)//2
                if nums[m] <= target:
                    l = m+1
                else:
                    r = m-1
            return r
        
        left,right = binarySearchLeft(),binarySearchRight()
        
        return [left,right] if left<=right else [-1,-1]
            
        
        
# @lc code=end

s = Solution()
print(s.searchRange([5, 7, 7, 8, 8, 10], 8))


